File: spec/models/playing_card_spec.rb
require "rails_helper"

RSpec.describe 'PlayingCard', type: :model do
  let(:card) { PlayingCard.new('Ace', 'Spades') }

  it 'returns rank' do
    expect(card.rank).to eq 'Ace'
  end

  it 'returns suit' do
    expect(card.suit).to eq 'Spades'
  end

  it 'returns value' do
    expect(card.value).to eq 13
  end

  it 'returns a string of first letter of rank and suit' do
    expect(card.toS).to eq 'sa'
  end

  describe 'json conversion' do
    let(:json_data) { card.as_json }
    let(:expanded_json) { PlayingCard.from_json(json_data) }

    it 'returns a json object of card' do
      expect(json_data['rank']).to eq('Ace')
      expect(json_data['suit']).to eq('Spades')
    end

    it 'changes the json card object back into card' do
      expect(expanded_json.rank).to eq('Ace')
      expect(expanded_json.suit).to eq('Spades')
    end
  end
end

File: spec/javascript/PlayingCardSpec.js
import React from 'react';
import ReactDOM from 'react-dom';
import PlayingCard from '../../app/javascript/models/PlayingCard.js'


describe("PlayingCard", () => {
  let rank, suit, card

  beforeEach(() => {
    rank = 'Ace';
    suit = 'Spades';
    card = new PlayingCard(rank, suit);
  });

  describe("rank", () => {
    it("should have a rank", () => {
      expect(card.rank()).toEqual(rank);
    });
  });

  describe("suit", () => {
    it("should have a suit", () => {
      expect(card.suit()).toEqual(suit);
    });
  });

  describe("value", () => {
    it("should have a value", function() {
      expect(card.value("Ace")).toEqual(13);
      expect(card.value("King")).toEqual(12);
    });
  });

  describe("toS", () => {
    it("should convert a card into a string representation", function() {
      expect(card.toS()).toEqual('sa');
      expect(card.toS()).not.toEqual('SA');
    });
  });

  describe("key", () => {
    it("should make a key of rank and suit", () => {
      expect(card.key()).toEqual("Ace-Spades");
    });
  });
});
